core/domain/learner_progress_services.py

```diff
@@ -0,0 +1,231 @@
+    if not activities_completed_model:
+        activities_completed_model = (
+            user_models.ActivitiesCompletedByLearnerModel(id=user_id))
+
+    exploration_created_ids = (
+        subscription_services.get_exploration_ids_subscribed_to(user_id))
+
+    if (exp_id not in exploration_created_ids and
+            exp_id not in activities_completed_model.completed_exp_ids):
+        # Remove the exploration from the in progress list (if present) as it is
+        # now completed.
+        remove_exp_from_incomplete_list(user_id, exp_id)
+        activities_completed_model.completed_exp_ids.append(exp_id)
+        activities_completed_model.put()
+
+def add_collection_id_to_completed_list(user_id, collection_id):
+    """Adds the collection id to the list of collections completed by the user.
+
+    Callers of this function must ensure that the user id and collection id are
+    valid.
+    """
+    activities_completed_model = (
+        user_models.ActivitiesCompletedByLearnerModel.get(
+            user_id, strict=False))
+    if not activities_completed_model:
+        activities_completed_model = (
+            user_models.ActivitiesCompletedByLearnerModel(id=user_id))
+
+    collection_created_ids = (
+        subscription_services.get_collection_ids_subscribed_to(user_id))
+
+    if (collection_id not in collection_created_ids and
+            collection_id not in
+            activities_completed_model.completed_collection_ids):
+        # Remove the collection from the in progress list (if present) as it is
+        # now completed.
+        remove_collection_from_incomplete_list(user_id, collection_id)
+        activities_completed_model.completed_collection_ids.append(
+            collection_id)
+        activities_completed_model.put()
+
+def add_exp_to_incomplete_list(
+        user_id, exploration_id, timestamp, state_name, version):
+    """Adds the exploration id to the incomplete list of the user.
+
+    Callers of this function must ensure that all the arguments provided are
+    valid.
+    """
+    incomplete_exp_model = (
+        user_models.IncompleteExplorationsModel.get(
+            user_id, strict=False))
+    if not incomplete_exp_model:
+        incomplete_exp_model = (
+            user_models.IncompleteExplorationsModel(
+                id=user_id))
+
+    completed_exp_ids = get_all_completed_exp_ids(user_id)
+
+    exploration_created_ids = (
+        subscription_services.get_exploration_ids_subscribed_to(user_id))
+
+    incomplete_exp_details = {
+        'timestamp': str(timestamp),
+        'state_name': state_name,
+        'version': version
+    }
+
+    incomplete_exp = {
+        exploration_id: incomplete_exp_details
+    }
+
+    exp_already_present = False
+    if (exploration_id not in completed_exp_ids and
+            exploration_id not in exploration_created_ids):
+
+        for exp in incomplete_exp_model.incomplete_exps:
+            if exploration_id == exp.keys()[0]:
```

Comment:
No idea what exp.keys()[0] is, just from reading this. Might want a better naming system. Also unclear how you guarantee the default sort order for a dict's keys.